File: app/src/main/java/com/radio/app/adapters/StationAdapter.java
package com.radio.app.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import com.radio.app.R;
import com.radio.app.models.RadioStation;

import java.util.List;

public class StationAdapter extends RecyclerView.Adapter<StationAdapter.ViewHolder> {
    private final Context ctx;
    private final List<RadioStation> stations;
    private final OnStationClickListener listener;

    public interface OnStationClickListener { void onStationClick(RadioStation s); void onStationLongClick(RadioStation s); }

    public StationAdapter(Context ctx, List<RadioStation> stations, OnStationClickListener listener) { this.ctx = ctx; this.stations = stations; this.listener = listener; }

    @NonNull @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) { return new ViewHolder(LayoutInflater.from(ctx).inflate(R.layout.item_station, parent, false)); }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder h, int pos) {
        RadioStation s = stations.get(pos);
        h.tvName.setText(s.getName());
        h.tvProgram.setText(s.getCurrentProgram());
        h.tvCountry.setText(s.getCountry());
        h.tvBitrate.setText(s.getBitrate() > 0 ? s.getBitrate() + " kbps" : "");
        h.ivCover.setImageResource(R.drawable.ic_radio);
        h.card.setOnClickListener(v -> { if (listener != null) listener.onStationClick(s); });
        h.card.setOnLongClickListener(v -> { if (listener != null) listener.onStationLongClick(s); return true; });
    }

    @Override public int getItemCount() { return stations.size(); }

    static class ViewHolder extends RecyclerView.ViewHolder {
        CardView card; ImageView ivCover; TextView tvName, tvProgram, tvCountry, tvBitrate;
        ViewHolder(@NonNull View v) { super(v); card = v.findViewById(R.id.card_view); ivCover = v.findViewById(R.id.iv_cover); tvName = v.findViewById(R.id.tv_name); tvProgram = v.findViewById(R.id.tv_program); tvCountry = v.findViewById(R.id.tv_country); tvBitrate = v.findViewById(R.id.tv_bitrate); }
    }
}

File: app/src/main/java/com/radio/app/adapters/StationAdapter.kt
package com.radio.app.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import com.radio.app.R
import com.radio.app.models.RadioStation

class StationAdapter(
    private val ctx: Context,
    private val stations: List<RadioStation>,
    private val listener: OnStationClickListener?
) : RecyclerView.Adapter<StationAdapter.ViewHolder>() {

    fun interface OnStationClickListener {
        fun onStationClick(s: RadioStation)
        fun onStationLongClick(s: RadioStation)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(ctx).inflate(R.layout.item_station, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val station = stations[position]
        holder.tvName.text = station.name
        holder.tvProgram.text = station.currentProgram
        holder.tvCountry.text = station.country
        holder.tvBitrate.text = if (station.bitrate > 0) "${station.bitrate} kbps" else ""
        holder.ivCover.setImageResource(R.drawable.ic_radio)
        holder.card.setOnClickListener { listener?.onStationClick(station) }
        holder.card.setOnLongClickListener {
            listener?.onStationLongClick(station)
            true
        }
    }

    override fun getItemCount(): Int = stations.size

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val card: CardView = view.findViewById(R.id.card_view)
        val ivCover: ImageView = view.findViewById(R.id.iv_cover)
        val tvName: TextView = view.findViewById(R.id.tv_name)
        val tvProgram: TextView = view.findViewById(R.id.tv_program)
        val tvCountry: TextView = view.findViewById(R.id.tv_country)
        val tvBitrate: TextView = view.findViewById(R.id.tv_bitrate)
    }
}
